File: Project Netwerken/src/NodeJS/global/model/command/VectorClock.js
const varint = require('../../utils/varint/varint');

class VectorClock {

    constructor(options) {
        if (options.buffer) {
            let size = varint.decode(options.buffer);
            this.array = [size];
            for (let i = 0; i < size; i++) {
                this.array[i] = varint.decode(options.buffer);
            }
        } else {
            this.array = [];
            for(let i = 0; i < options.size; i++){
                this.array[i] = 0;
            }
        }
    }

/*    // OLD CONSTRUCTOR
    constructor(buffer){
        let size = varint.decode(buffer);
        this.array = [size];
        for (let i = 0; i < size; i++) {
            this.array[i] = varint.decode(buffer);
        }
    }

    // OLD CONSTRUCTOR
    constructor(size) {
        this.array = [];
        for(let i = 0; i < size; i++){
            this.array[i] = 0;
        }
    }*/

    write(buffer){
        buffer.write(varint.encode(this.array.length));
        for(let i = 0; i < this.array.length; i++){
            buffer.write(varint.encode(this.array[i]));
        }
    }
}

module.exports = VectorClock;

File: Project Netwerken/src/Global/src/model/command/VectorClock.java
package model.command;

import utils.VarInt;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;

/**
 * With this vectorclock you can check if you have received the last message from every node.
 */
public class VectorClock {
    private int[] array;

    public VectorClock(ByteBuffer in) {
        int size = VarInt.readUnsignedVarInt(in);
        array = new int[size];
        for (int i = 0; i < size; i++) {
            array[i] = VarInt.readUnsignedVarInt(in);
        }
    }

    public VectorClock(int size) {
        assert size > 1: "size must be bigger than one";
        this.array = new int[size];
    }

    public int[] getArray() {
        return array;
    }

    public int getSize() {
        return array.length;
    }

    public int incrementVector(int nodeId){
        return ++array[nodeId];
    }

    public void write(ByteArrayOutputStream out) throws IOException {
        out.write(VarInt.writeUnsignedVarInt(array.length));
        for (int vector : array) {
            out.write(VarInt.writeUnsignedVarInt(vector));
        }
    }
}
